### PR TITLE
Exporter bug & Exporter Refactor 

### PR DESCRIPTION
**Pull request by matthewmcnew**
Fixes buildpack/roadmap#34

### CODE REVIEW

File: exporter.go [hunk 1]
```diff
@@ -142,119 +39,152 @@ func (e *Exporter) exportImage(layersDir, appDir, launcher string, runImage, ori
...
+		return errors.Wrap(err, "exporting launcher layer")
+	}
 
-		layerKeys := make([]string, 0, len(bp.Layers))
-		for n, _ := range bp.Layers {
-			layerKeys = append(layerKeys, n)
+	for _, bp := range e.Buildpacks {
+		bpDir, err := readBuildpackLayersDir(layersDir, bp.EscapedID())
+		if err != nil {
+			panic(err)
```
Comment (sclevine) [inline exporter.go]: remove panic

Comment (ekcasey) [reply exporter.go]: fixed it

File: exporter.go [hunk 2]
```diff
@@ -142,119 +39,152 @@ func (e *Exporter) exportImage(layersDir, appDir, launcher string, runImage, ori
...
-	if prevBP == nil {
-		return LayerMetadata{}
+func (e *Exporter) addOrReuseLayer(image *loggingImage, layer identifiableLayer, previousSha string) (string, error) {
+	sha, err := e.exportTar(layer.Path())
+	if err != nil {
+		return "", errors.Wrapf(err, "exporting layer '%s'", layer.Identifier())
+	}
+	if sha == previousSha {
+		return sha, image.ReuseLayer(layer.Identifier(), previousSha)
+	} else {
```
Comment (sclevine) [inline exporter.go]: Could remove else

File: layers.go [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+
+func (bp *bpLayer) read() (LayerMetadata, error) {
+	var metadata LayerMetadata
+	tomlPath := bp.path + ".toml"
+	fh, err := os.Open(tomlPath)
+	if err != nil {
+		return LayerMetadata{}, err
+	}
+	defer fh.Close()
+	_, err = toml.DecodeFile(tomlPath, &metadata)
```
Comment (sclevine) [inline layers.go]: could inline

File: layers.go [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+	if err != nil {
+		return LayerMetadata{}, err
+	}
+	if sha, err := ioutil.ReadFile(bp.path + ".sha"); err == nil {
+		metadata.SHA = string(sha)
```
Comment (sclevine) [inline layers.go]: Don't invert error flow

File: layers.go [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+	if err := os.Remove(bp.path + ".toml"); err != nil && !os.IsNotExist(err) {
+		return err
+	}
+	return nil
+}
+
+func (bp *bpLayer) writeMetadata(metadataLayers map[string]LayerMetadata) error {
+	layerMetadata := metadataLayers[bp.name()]
+	path := filepath.Join(bp.path + ".toml")
+	if err := os.MkdirAll(filepath.Dir(path), 0755); err != nil {
```
Comment (sclevine) [inline layers.go]: `0755` - > `0777`

File: layers.go [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+}
+
+func (bp *bpLayer) hasLocalContents() bool {
+	_, err := ioutil.ReadDir(bp.path)
+
+	return !os.IsNotExist(err)
+}
+
+func (bp *bpLayer) writeSha(sha string) error {
+	if err := ioutil.WriteFile(filepath.Join(bp.path+".sha"), []byte(sha), 0755); err != nil {
```
Comment (sclevine) [inline layers.go]: `0777`

File: logging_image.go [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+func (li *loggingImage) SetEntrypoint(entryPoint string) error {
+	li.Out.Printf("setting entrypoint '%s'\n", entryPoint)
+	return li.image.SetEntrypoint(entryPoint)
+}
+
+func (li *loggingImage) SetEmptyCmd() error {
+	li.Out.Println("setting empty cmd")
+	return li.image.SetCmd()
+
```
Comment (sclevine) [inline logging_image.go]: remove newline

**Pull request merged**